doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,810 @@
+        decli [label="Code\nStatement: Declaration\nType: unsigned\nName: i.1\nValue: 1",shape=rectangle]
+        cond [label="Expression: <=",shape=rectangle]
+        condlhs [label="Expression: Variable i.1",shape=rectangle]
+        condrhs [label="Expression: Variable n.1",shape=rectangle]
+        inc [label="Code\nExpression: ++ - postfix",shape=rectangle]
+        incarg [label="Expression: Variable i.1",shape=rectangle]
+        body [label="Code\nStatement: Block"]
+        mult [label="Code\nExpression: *=",shape=rectangle]
+        multlhs [label="Expression: Variable fac.1",shape=rectangle]
+        multrhs [label="Expression: Variable i.1",shape=rectangle]
+        return [label="Code\nStatement: return"]
+        returnexpr [label="Expression: Variable fac.1",shape=rectangle]
+        factorial -> {declfac, for, return}
+        for -> {decli,cond,inc,body}
+        cond -> {condlhs,condrhs}
+        inc -> incarg
+        body -> mult -> {multlhs, multrhs}
+        return -> returnexpr
+        globalfactorial -> {parameters,factorial}
+        parameters -> parameter
+
+        globalmain [label="Global - code\nType: int\nName: main"]
+        parametersmain [label="Parameters"]
+        parameterargc [label="Parameter\nType: int\nName: argc.1"]
+        parameterargv [label="Parameter\nType: const char **\nName: argv.1"]
+        mainbody [label="Code\nStatement: Block"]
+        main0 [label="Code\nStatement: Declaration\nType: unsigned\nName: n.2\nValue: Expression"]
+        main1 [label="Code\nExpression: Call"]
+        main1fun [label="Expression: Global atoi"]
+        main1arg [label="Arguments"]
+        main1arg1 [label="Expression: Index"]
+        main1arg11 [label="Expression: Variable argv.1"]
+        main1arg12 [label="Expression: 1"]
+        main2 [label="Code\nExpression: Call"]
+        main2fun [label="Expression: Global printf"]
+        main2arg [label="Arguments"]
+        main2arg1 [label="Expression: ..."]
+        main2arg2 [label="Expression: Variable n.2"]
+        main2arg3 [label="Expression: Call"]
+        main2arg3fun [label="Expression: Global factorial"]
+        main2arg3arg [label="Arguments"]
+        main2arg3arg1 [label="Expression: Variable n.2"]
+        main3 [label="Code\nStatement: return"]
+        main3expr [label="Expression: 0"]
+        globalmain -> {parametersmain,mainbody}
+        parametersmain -> {parameterargc, parameterargv}
+        mainbody -> {main0,main1,main2}
+        main1 -> {main1fun, main1arg}
+        main1arg -> main1arg1 -> {main1arg11,main1arg12}
+        main2 -> {main2fun, main2arg}
+        main2arg -> {main2arg1,main2arg2,main2arg3}
+        main2arg3-> {main2arg3fun, main2arg3arg}
+        main2arg3arg -> main2arg3arg1
+        main3 -> mainexpr
+}
+\enddot
+
+Note that the parameter `n` in `factorial` and the local variable `n` in
+`main` are now disambiguated as `n.1` and `n.2`; furthermore, we leave
```

Comment:
CBMC prefixes it with the method name:: and further numbering to distinguish scopes